### PR TITLE
Use separate searchers for "search visibility" vs "move indexing buffer to disk

### LINKED ISSUE #15768: Use separate searchers for "search visibility" vs "move indexing buffer to disk"
Spinoff from #14121...

Today, when ES detects it's using too much heap vs the configured indexing buffer (default 10% of JVM heap) it opens a new searcher to force Lucene to move the bytes to disk, clear version map, etc.

But this has the unexpected side effect of making newly indexed/deleted documents visible to future searches, which is not nice for users who are trying to prevent that, e.g. #3593.

As @uschindler suggested in that issue, I think ES should have two separate searchers from the engine: one for search visibility, only ever refreshed according to the user's wishes, and another, used internally for freeing up heap, version map lookups, etc.  Lucene will be efficient about this, sharing segment readers across those two searchers.

I haven't started on this (need to finish #14121 first!) so if someone wants to take it, please feel free!

### LINKED ISSUE #26912: Explore using a engine-private IndexReader for realtime-get etc.
This is a spinoff from #26802 where we potentially pay a big price on rebuilding caches etc.  when updates / realtime-get is used. We are refreshing the internal reader for realtime gets which causes for instance global ords to be rebuild. I think we can gain quite a bit if we'd use a reader that is only used for GETs and not for searches etc. that way we can also solve problems of searchers being refreshed unexpectedly aside of replica recovery / relocation.

### PR DESCRIPTION
**Pull request by s1monw**
Today, when ES detects it's using too much heap vs the configured indexing
buffer (default 10% of JVM heap) it opens a new searcher to force Lucene to move
the bytes to disk, clear version map, etc.

But this has the unexpected side effect of making newly indexed/deleted
documents visible to future searches, which is not nice for users who are trying
to prevent that, e.g. #3593.

This is also an indirect spinoff from #26802 where we potentially pay a big
price on rebuilding caches etc. when updates / realtime-get is used. We are
refreshing the internal reader for realtime gets which causes for instance
global ords to be rebuild. I think we can gain quite a bit if we'd use a reader
that is only used for GETs and not for searches etc. that way we can also solve
problems of searchers being refreshed unexpectedly aside of replica recovery /
relocation.

Closes #15768
Closes #26912

### CODE REVIEW

**Comment by uschindler** [comment]
Having 2 IndexReaders (in fact DirectoryReaders) open at same time will not have any impact on memory usage, as both share the same LeafReaders. The memory of stale segments is only freed if the one with oldest/nolonger uptodate segments is released.

@s1monw was this your question so you pinged me?

**Comment by s1monw** [comment]
@uschindler thanks, I was aware of this that's why I added it. I pinged you since you where involved in the original issue here #15768

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: shall we make the refreshExternal an EnumSet of SearcherScope?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ++

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I wonder if we can now always use this "refresh" path here rather than the `IndexWriter.flush` - refresh is now much lighter? (not sure what other side effects are there).

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Doesn't the comment about releasing segments in tryRenewSyncCommit holds here too?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Isn't tryRenewSyncCommit already doing it in this case?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: same comment - if `tryRenewSyncCommit` returned true, it already refreshed, no?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: nit - should we use try-with-resources to avoid leaking on failure?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks for picking this up so quickly. I did a pass and left mostly nits around naming (sorry :)). I did leave some comments about what seems like excessive refreshes which is why I marked this as request changes.

Also - with the double searcher managers and their relationship to the version map I got a bit worried about the IndexinMemoryController calling `writeIndexingBuffer`. I couldn't find any test for that. Do we have one? I think we should have a concurrent read/write by ID engine unit test which validates it's safe to call refresh/flush/writeIndexingBuffer without us messing up. WDYT?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: +1 I think that's good

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think I did some copy past mess here. I will fix

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: see my comment above sorry for the noise

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: will do

**Comment by s1monw** [comment]
@bleskes I am not sure if you love or hate this 🤷‍♂️ I took it one step further and solely use that internal searcher everywhere. the external searcher is used for non-RT get and search but internally it's all the done via the private reader. This also means we don't refresh anymore during `force_merge` and `flush` but to keep it BWC I moved these calls into `IndexShard`. Conceptually I think we should not refresh in these situations and we might change that in `7.0` another option is to make refresh optional on these API and change a default in 7 so folks can choose.

I personally think this is much cleaner now.

**Comment by mikemccand** [comment]
Yay!!

**Comment by bleskes** [comment]
@mikemccand you're alive!

**Review by uschindler — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsResponse.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsResponse.java]: :) it's a good sign you bumped into this, isn't this changing semantics -  before when there were no docs, we won't output the Field Stats sub object

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: with the move to internal external (rather then get/search), I think this method can go and that also means that the Engine.acquireSearch(source, scope) can be made protected.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ++ sneaky and nice.  I think we can extend InternalEngineTests.testSimpleOperations with:

```
        // but, we can still get it (in realtime)
        getResult = engine.get(newGet(true, doc), searcherFactory);
        assertThat(getResult.exists(), equalTo(true));
        assertThat(getResult.docIdAndVersion(), notNullValue());
        getResult.release();

+        // but not real time is not yet visible
+        getResult = engine.get(newGet(false, doc), searcherFactory);
+        assertThat(getResult.exists(), equalTo(false));
+        getResult.release();

````

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Never mind. I see it now :(

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I don't see the fall through?

File: core/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java]: ++

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: OK. Now I see why the engine changes can be done in 6.x. I would opt for keeping as it was (i.e. full refresh on flush) and do as a small following in 7.0 to change the refresh command in the engine + mark it as breaking etc. Just a suggestion.

**Review by bleskes — APPROVED** [review]
This is great ❤️ . I left suggestions but there is no need for another review.

**Comment by s1monw** [comment]
@uschindler @bleskes thanks for the reviews I will wait for CI and merge it

File: core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsResponse.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsResponse.java]: expand the diff below and you will see all the void 💃

**Comment by uschindler** [comment]
Thanks @s1monw !

**Pull request merged**
